src/NUnitEngine/nunit.engine/Services/ExtensionService.cs

```diff
@@ -0,0 +1,154 @@
+
+            var ep = GetExtensionPoint(typeof(T));
+            foreach (var node in ep.Extensions)
+                extensions.Add((T)node.ExtensionObject);
+
+            return extensions.ToArray();
+        }
+
+        #region Service Overrides
+
+        public override void StartService()
+        {
+            var thisAssembly = Assembly.GetExecutingAssembly();
+            var startDir = new DirectoryInfo(AssemblyHelper.GetDirectoryName(thisAssembly));
+
+            FindExtensionPoints(thisAssembly);
+            FindExtensionsInDirectory(startDir);
+
+            base.StartService();
+        }
+
+        private void FindExtensionPoints(Assembly assembly)
+        {
+            foreach (ExtensionPointAttribute attr in assembly.GetCustomAttributes(typeof(ExtensionPointAttribute), false))
+            {
+                var ep = new ExtensionPoint(attr.Path, attr.Type)
+                {
+                    Description = attr.Description
+                };
+
+                _extensionPoints.Add(ep);
+
+                // TODO: These Adds will throw if the path or type is a duplicate.
+                // We should give an error message instead.
+                _pathIndex.Add(ep.Path, ep);
+                _typeIndex.Add(ep.Type, ep);
+            }
+        }
+
+        private void FindExtensionsInDirectory(DirectoryInfo startDir)
+        {
+            var addinsFiles = startDir.GetFiles("*.addins");
+            if (addinsFiles.Length > 0)
+                foreach (var file in addinsFiles)
+                    ProcessAddinsFile(startDir.FullName, file.FullName);
+            else
+                foreach (var file in startDir.GetFiles("*.dll"))
+                    FindExtensionsInAssembly(file.FullName);
+        }
+
+        private void ProcessAddinsFile(string baseDir, string fileName)
+        {
+            var doc = new XmlDocument();
+
+            using (var rdr = new StreamReader(fileName))
+            {
+                doc.Load(rdr);
+                foreach (XmlNode dirNode in doc.SelectNodes("Addins/Directory"))
+                {
+                    var path = Path.Combine(baseDir, dirNode.InnerText);
+                    FindExtensionsInDirectory(new DirectoryInfo(path));
+                }
+            }
+        }
+
+        private void FindExtensionsInAssembly(string assemblyName)
+        {
+            var module = AssemblyDefinition.ReadAssembly(assemblyName).MainModule;
+            foreach (var type in module.GetTypes())
+                foreach (var attr in type.CustomAttributes)
+                    if (attr.AttributeType.FullName == "NUnit.Engine.Extensibility.ExtensionAttribute")
+                    {
+                        var node = new ExtensionNode(assemblyName, type.FullName);
+                        node.Path = (string)attr.ConstructorArguments[0].Value;
+                        _extensions.Add(node);
+
+                        var ep = GetExtensionPoint(node.Path);
```

Comment:
How would you deal with duplicate interfaces? I think we can minimize the use of the path, but not eliminate it unless we say that no two extension points can accept the same interface. That seems arbitrary and could cause trouble down the line.